// @amuzlab/fs-promise/file-system/watcher/AbstractFile.js
'use strict'

const path = require('path')

class AbstractFile extends require('events') {
    constructor (filePath, fileName, stat, relativePath, option) {
        super()

        Object.defineProperties(
            this,
            {
                _filePath: {
                    enumerable: true,
                    writable: true,
                    value: filePath
                },
                _fileName: {
                    enumerable: true,
                    writable: true,
                    value: fileName
                },
                _stat: {
                    writable: true,
                    value: stat
                },
                _relativePath: {
                    writable: true,
                    value: relativePath
                },
                _option: {
                    writable: true,
                    value: option
                }
            })
    }

    get fileName () {
        return this._fileName
    }

    get filePath () {
        return this._filePath
    }

    get fullPath () {
        return path.join(this.filePath, this.fileName)
    }

    get size () {
        return this.stat ? this.stat.size : 0
    }

    get stat () {
        return this._stat
    }

    get isDirectory () {
        return false
    }

    get relativePath () {
        return this._relativePath
    }

    get option () {
        return this._option
    }

    isSelf (filePath, fileName) {
        filePath instanceof File && (filePath = filePath.filePath, fileName = filePath.fileName)

        return this.filePath === filePath && this.fileName === fileName
    }
}

module.exports = exports = AbstractFile
