// @amuzlab/utils/index.js
'use strict'

const object = require('./object')

Object.assign(exports, require('./util'))
object.defineProperties(
    exports,
    {
        /**
         * @variation   global
         * @global
         * @name        object
         * @desc        object utility module
         * @type        {module:object}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        object: require('./object'),
        /**
         * @variation   global
         * @global
         * @name        date
         * @desc        date utility module
         * @type        {module:date}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        date: require('./date'),
        /**
         * @variation   global
         * @global
         * @name        string
         * @desc        string utility module
         * @type        {module:string}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        string: require('./string'),
        /**
         * @variation   global
         * @global
         * @name        path
         * @desc        path utility module
         * @type        {module:path}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        path: require('./path'),
        /**
         * @variation   global
         * @global
         * @name        enum
         * @desc        enum utility module
         * @type        {module:enum}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        enum: require('./enum'),
        /**
         * @variation   global
         * @global
         * @name        validation
         * @desc        validation utility module
         * @type        {module:validation}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        validation: require('./validation'),
        /**
         * @variation   global
         * @global
         * @name        async
         * @desc        async utility module
         * @type        {module:async}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        async: require('./async'),
        /**
         * @variation   global
         * @global
         * @name        Timer
         * @desc        Timer utility class
         * @type        {Timer}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        Timer: require('./Timer'),
        /**
         * @variation   global
         * @global
         * @name        event
         * @desc        event utility class
         * @type        {EventEmitter}
         * @static
         * @public
         * @readonly
         * @since       0.0.1
        */
        event: require('./EventEmitter'),
        /**
         * @variation   global
         * @global
         * @name        unit
         * @desc        unit utility module
         * @type        {module:unit}
         * @static
         * @public
         * @readonly
         * @since       1.2.6
        */
        unit: require('./unit')
    },
    { enumerable: true })
